mlir/lib/Analysis/Presburger/PresburgerSet.cpp

```diff
@@ -395,39 +395,170 @@ Optional<uint64_t> PresburgerSet::computeVolume() const {
+  for (unsigned l = 0, e = A.getNumEqualities(); l < e; ++l)
+    only_redundant_eqsA &= simplex2.isRedundantEquality(A.getEquality(l));
+
+  bool only_redundant_eqsB = true;
+  for (unsigned l = 0, e = B.getNumEqualities(); l < e; ++l)
+    only_redundant_eqsB &= simplex1.isRedundantEquality(B.getEquality(l));
+
+  SmallVector<ArrayRef<int64_t>, 2> redundant_ineqsA;
+  SmallVector<ArrayRef<int64_t>, 2> cutting_ineqsA;
+
+  SmallVector<ArrayRef<int64_t>, 2> redundant_ineqsB;
+  SmallVector<ArrayRef<int64_t>, 2> cutting_ineqsB;
+
+  for (unsigned l = 0, e = A.getNumInequalities(); l < e; ++l) {
+    Simplex::IneqType type = simplex2.ineqType(A.getInequality(l));
+    if (type == Simplex::IneqType::Redundant) {
+      redundant_ineqsA.push_back(A.getInequality(l));
+    } else if (type == Simplex::IneqType::Cut) {
+      cutting_ineqsA.push_back(A.getInequality(l));
+    } else {
+      return false;
     }
+  }
+  for (unsigned l = 0, e = B.getNumInequalities(); l < e; ++l) {
+    Simplex::IneqType type = simplex1.ineqType(B.getInequality(l));
+    if (type == Simplex::IneqType::Redundant) {
+      redundant_ineqsB.push_back(B.getInequality(l));
+    } else if (type == Simplex::IneqType::Cut) {
+      cutting_ineqsB.push_back(B.getInequality(l));
+    } else {
+      return false;
+    }
+  }
+
+  if (cutting_ineqsA.size() == 0 && only_redundant_eqsA) {
+    addCoalescedPolyhedron(polyhedrons, i, j, A);
+    return true;
+  }
+
+  if (cutting_ineqsB.size() == 0 && only_redundant_eqsB) {
+    addCoalescedPolyhedron(polyhedrons, i, j, B);
+    return true;
+  }
+
+  if (cutCase(polyhedrons, i, j, redundant_ineqsA, cutting_ineqsA,
+              redundant_ineqsB, cutting_ineqsB)) {
+    return true;
+  }
 
-    // Check whether `IntegerPolyhedron[i]` is contained in any Poly, that is
-    // different from itself and not yet marked as redundant.
-    for (unsigned j = 0, e = integerPolyhedrons.size(); j < e; ++j) {
-      if (j == i || isRedundant[j])
-        continue;
+  if (cutCase(polyhedrons, j, i, redundant_ineqsB, cutting_ineqsB,
+              redundant_ineqsA, cutting_ineqsA)) {
+    return true;
+  }
+
+  return false;
+}
 
-      if (simplex.isRationalSubsetOf(integerPolyhedrons[j])) {
-        isRedundant[i] = true;
+PresburgerSet PresburgerSet::coalesce() const {
+  PresburgerSet newSet =
+      PresburgerSet::getEmptySet(getNumDimIds(), getNumSymbolIds());
+  llvm::SmallBitVector toDelete(getNumPolys());
+  SmallVector<IntegerPolyhedron, 2> polyhedrons = integerPolyhedrons;
+
+  for (unsigned i = 0; i < polyhedrons.size(); ++i) {
+    for (unsigned j = i + 1; j < polyhedrons.size(); ++j) {
+      bool res_status = coalescePair(i, j, polyhedrons);
+      if (res_status) {
+        i--;
         break;
       }
     }
   }
 
-  for (unsigned i = 0, e = integerPolyhedrons.size(); i < e; ++i)
-    if (!isRedundant[i])
-      newSet.unionPolyInPlace(integerPolyhedrons[i]);
+  if (polyhedrons.size() == 1) {
```

Comment:
This is a not very nice workaround to handle cases for which all input sets are empty. I have such a case as a test, so I decided to do it here too.